api/src/main/java/io/opencensus/trace/MessageEvent.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Timestamp;
+import io.opencensus.internal.BaseMessageEvent;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A class that represents a generic messaging event. Unlike {@link NetworkEvent} recorded by RPC
+ * client or server, this class can represent messaging happened in any layer, especailly higher
+ * application layer. Thus, it can be used when recording events in pipeline works, in-process
+ * bidirectional streams and batch processing.
+ *
+ * <p>It requires a {@link Type type} and a message id that serves to uniquely identify each
+ * message. It can optionally have information about the kernel time and message size.
+ *
+ * @since 0.12.0
+ */
+@Immutable
+@AutoValue
+// Suppress Checker Framework warning about missing @Nullable in generated equals method.
+@AutoValue.CopyAnnotations
+@SuppressWarnings("nullness")
+public abstract class MessageEvent extends BaseMessageEvent {
+  /** Available types for a {@code MessageEvent}. */
+  public enum Type {
+    /** When the message was sent. */
+    SENT,
+    /** When the message was received. */
+    RECEIVED,
+  }
+
+  /**
+   * Returns a new {@link Builder} with default values.
+   *
+   * @param type designates whether this is a send or receive message.
+   * @param messageId serves to uniquely identify each message.
+   * @return a new {@code Builder} with default values.
+   * @throws NullPointerException if {@code type} is {@code null}.
+   */
+  public static Builder builder(Type type, long messageId) {
+    return new AutoValue_MessageEvent.Builder()
+        .setType(checkNotNull(type, "type"))
+        .setMessageId(messageId)
+        // We need to set a value for the message size because the autovalue requires all
+        // primitives to be initialized.
+        .setUncompressedMessageSize(0)
+        .setCompressedMessageSize(0);
+  }
+
+  /**
+   * Returns the kernel timestamp associated with the {@code MessageEvent} or {@code null} if not
+   * set.
+   *
+   * @return the kernel timestamp associated with the {@code MessageEvent} or {@code null} if not
+   *     set.
+   */
+  @Nullable
+  public abstract Timestamp getKernelTimestamp();
```

Comment:
We do not have this in the MessageEvent.

Reply:
@bogdandrutu if users have already been using `NetworkEvent.getKernelTimestamp()` to record kernelTimeStamp, what should they use to do so when switching to `MessageEvent`? Or they should just drop that field?

Reply:
Nobody uses it. Dropping is fine.